src/main/host/descriptor/eventfd.rs

```diff
@@ -0,0 +1,224 @@
+    ) -> SyscallResult
+    where
+        W: std::io::Write + std::io::Seek,
+    {
+        // eventfds don't support seeking
+        if offset != 0 {
+            return Err(Errno::ESPIPE.into());
+        }
+
+        // eventfd(2): "Each successful read(2) returns an 8-byte integer"
+        const NUM_BYTES: usize = 8;
+
+        // this check doesn't guarentee that we can write all bytes since the stream length is only
+        // a hint
+        if usize::try_from(bytes.stream_len_bp()?).unwrap() < NUM_BYTES {
+            log::trace!(
+                "Reading from eventfd requires a buffer of at least {} bytes",
+                NUM_BYTES
+            );
+            return Err(Errno::EINVAL.into());
+        }
+
+        if self.counter == 0 {
+            log::trace!("Eventfd counter is 0 and cannot be read right now");
+            return Err(Errno::EWOULDBLOCK.into());
+        }
+
+        // behavior defined in `man 2 eventfd`
+        if self.semaphore_mode {
+            const ONE: [u8; NUM_BYTES] = 1u64.to_ne_bytes();
+            bytes.write_all(&ONE)?;
+            self.counter -= 1;
+        } else {
+            let to_write: [u8; NUM_BYTES] = self.counter.to_ne_bytes();
+            bytes.write_all(&to_write)?;
+            self.counter = 0;
+        }
+
+        self.update_state(event_queue);
+
+        return Ok(NUM_BYTES.into());
+    }
+
+    pub fn write<R>(
+        &mut self,
+        mut bytes: R,
+        offset: libc::off_t,
+        event_queue: &mut EventQueue,
+    ) -> SyscallResult
+    where
+        R: std::io::Read + std::io::Seek,
+    {
+        // eventfds don't support seeking
+        if offset != 0 {
+            return Err(Errno::ESPIPE.into());
+        }
+
+        // eventfd(2): "A write(2) call adds the 8-byte integer value supplied in its buffer to the
+        // counter"
+        const NUM_BYTES: usize = 8;
+
+        // this check doesn't guarentee that we can read all bytes since the stream length is only
+        // a hint
+        if usize::try_from(bytes.stream_len_bp()?).unwrap() < NUM_BYTES {
+            log::trace!(
+                "Writing to eventfd requires a buffer with at least {} bytes",
+                NUM_BYTES
+            );
+            return Err(Errno::EINVAL.into());
+        }
+
+        let mut read_buf = [0u8; NUM_BYTES];
+        bytes.read_exact(&mut read_buf)?;
+        let value: u64 = u64::from_ne_bytes(read_buf);
+
+        if value == u64::MAX {
+            log::trace!("We do not allow writing the max counter value");
+            return Err(Errno::EINVAL.into());
+        }
+
+        const MAX_ALLOWED: u64 = u64::MAX - 1;
+        if value > MAX_ALLOWED - self.counter {
+            log::trace!("The write value does not currently fit into the counter");
+            return Err(Errno::EWOULDBLOCK.into());
+        }
+
+        self.counter += value;
+        self.update_state(event_queue);
+
+        return Ok(NUM_BYTES.into());
```

Comment:
Do we want the `return` statements (punctuated with `;`) throughout the read and write functions, or is that considered non-standard rust style?

Reply:
I'm not sure what you mean here. Since this is the last return statement of the function, I did change the `return Ok(NUM_BYTES.into());` here into just `Ok(NUM_BYTES.into()`.